Telcodocs 1771 Extend workload partitioning to support cpu limits (#75078)
openshift/openshift-docs: scalability_and_performance/enabling-workload-partitioning.adoc

toc::[]

> Workload partitioning segregates particular workloads into distinct CPU sets. The primary objective is ensure that the platform pods will not exceed a specified CPU usage limit. You can use workload partitioning to isolate {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs. This ensures that the remaining CPUs in their deployment remain untouched and available exclusively for their own workloads. The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).

Reviewer: It's not the workloads that get segregated but rather the compute node cpu resources.

And the objective is not to ensure that platform pods don't exceed usage limits but rather that system tasks are restricted to specific CPUs only. And also what you mention below, that the cpu capacity of the host is correctly accounted for

In the context of enabling workload partitioning and managing CPU resources effectively, nodes within the cluster must identify themselves as supporting a new resource type called `management.workload.openshift.io/cores` and accurately report their CPU capacity. Workload partitioning can be enabled during cluster installation only by adding the additional field: `cpuPartitioningMode` to the `install-config.yaml` file.

When workload partitioning is enabled, the `management.workload.openshift.io/cores` resource allows the scheduler to correctly assign pods based on the cpushares capacity of the host, not just the default cpuset. This ensures more precise allocation of resources for workload partitioning scenarios.

> Workload partitioning respects CPU requests and limits defined on pods. From 4.16 accurate CPU usage limitations are set for platform pods utilizing CPU partitioning. Resource requests, rather than limits, are the primary value in setting accurate CPU usage limitations. 

Reviewer 2: Maybe rework this paragraph a bit,

Something like this, maybe?

> Workload partitioning respects CPU requests and limits defined in the pod spec. As of 4.16 accurate CPU usage limits are set for platform pods utilizing CPU partitioning. Since workload partitioning uses the custom resource type of `management.workload.openshift.io/cores`,  the values for `requests` and `limits` will equal the same due to a requirement by Kubernetes for extended resources. However, the annotations modified by workload partitioning will correctly reflect the desired limits.

> If, for example, a request is set to 20 and the limit to 40, the scheduler prioritizes the lower value to allocate resources according to actual usage. Requesting 20 represents the typical need, while 40 is the upper limit. Prioritizing requests over limits optimizes resource allocation, aligning with usage needs and avoiding excess provisioning.

Reviewer: Is this only relevant to to platform Pods?
If so, why is this relevant to the end user, they are not supposed to create those

Reviewer 2: This might be too much detail, we're essentially describing how the scheduler works here. 

The only thing I would want to make mention of is something I wrote above regarding the custom resource values equaling the same for requests and limits. That's a requirement of Kubernetes

https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#consuming-extended-resources
> Note: Extended resources cannot be overcommitted, so request and limit must be equal if both are present in a container spec.

toc::[]

> Workload partitioning segregates compute node CPU resources into distinct CPU sets. The primary objective is to keep platform pods on the specified cores to avoid interrupting the CPUs the customer workloads are running on. Workload partitioning isolates {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs. This ensures that the remaining CPUs in their deployment remain untouched and available exclusively for their own workloads. The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).

Reviewer 2: Small nit for the `This ensures that the remaining CPUs in their deployment remain untouched and available exclusively for their own workloads.` sentence. Maybe something like the one below is a little clearer wdyt?

>This ensures that the remaining CPUs in the cluster deployment are untouched and available exclusively for non platform workloads.

> In the context of enabling workload partitioning and managing CPU resources effectively, nodes within the cluster must identify themselves as supporting a new resource type called `management.workload.openshift.io/cores` and accurately report their CPU capacity. Workload partitioning can be enabled during cluster installation only by adding the additional field: `cpuPartitioningMode` to the `install-config.yaml` file.

Reviewer 2: We should change this a bit so customers know what is being done automatically and what they might need to be aware of regarding the nodes.

Something like this maybe?

>In the context of enabling workload partitioning and managing CPU resources effectively, nodes that are not configured correctly will not be allowed to join the cluster via a node admission webhook. When the workload partitioning feature is enabled, the Machine Config Pools for `master` and `worker` will be supplied with configurations for nodes to use. Adding new nodes to these pools will make sure they are correctly configured before joining the cluster. 
>
>Currently, it is required that nodes are homogenous per Machine Config Pool in order to ensure correct CPU Affinity is set across all nodes with in that pool. Once admitted nodes within the cluster will identify themselves as supporting a new resource type called `management.workload.openshift.io/cores` and accurately report their CPU capacity. Workload partitioning can be enabled during cluster installation only by adding the additional field: `cpuPartitioningMode` to the `install-config.yaml` file.

toc::[]

> Workload partitioning segregates compute node CPU resources into distinct CPU sets. The primary objective is to keep platform pods on the specified cores to avoid interrupting the CPUs the customer workloads are running on. Workload partitioning isolates {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs. This ensures that the remaining CPUs in the cluster deployment are untouched and available exclusively for non platform workloads. The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).

Reviewer 3: * See ISG on avoiding "segregate". 
* This first paragraph felt pretty big, so suggest breaking it up somehow.
* Re: "non platform" - ISG seems to suggest no space or hyphen for the "non" prefix. Personally "nonplatform" feels weird to me. Maybe "non-platform"? Either way, something other than a space.
```suggestion
Workload partitioning separates compute node CPU resources into distinct CPU sets. The primary objective is to keep platform pods on the specified cores to avoid interrupting the CPUs the customer workloads are running on. 

Workload partitioning isolates {product-title} services, cluster management workloads, and infrastructure pods to run on a reserved set of CPUs. This ensures that the remaining CPUs in the cluster deployment are untouched and available exclusively for non-platform workloads. The minimum number of reserved CPUs required for the cluster management is four CPU Hyper-Threads (HTs).
```

> In the context of enabling workload partitioning and managing CPU resources effectively, nodes that are not configured correctly will not be permitted to join the cluster through a node admission webhook. When the workload partitioning feature is enabled, the Machine Config Pools for master and worker will be supplied with configurations for nodes to use. Adding new nodes to these pools will make sure they are correctly configured before joining the cluster.

Reviewer 3: * Re: "master", can it be "control plane" here? Per [SSG](https://redhat-documentation.github.io/supplementary-style-guide/#_master_and_slave) and [repo term glossary](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc#control-plane).
* I think "machine config pool" is mostly lowercased throughout the repo.

```suggestion
In the context of enabling workload partitioning and managing CPU resources effectively, nodes that are not configured correctly will not be permitted to join the cluster through a node admission webhook. When the workload partitioning feature is enabled, the machine config pools for control plane and worker will be supplied with configurations for nodes to use. Adding new nodes to these pools will make sure they are correctly configured before joining the cluster.
```

Workload partitioning ensures that CPU requests and limits specified in the pod's configuration are respected. From 4.16 accurate CPU usage limits are set for platform pods through CPU partitioning. As workload partitioning uses the custom resource type of `management.workload.openshift.io/cores`, the values for requests and limits will be the same due to a requirement by Kubernetes for extended resources. However, the annotations modified by workload partitioning will correctly reflect the desired limits. 

[NOTE]
====
Extended resources cannot be overcommitted, so request and limit must be equal if both are present in a container spec.
====

include::modules/enabling-workload-partitioning.adoc[leveloffset=+1]

include::modules/create-perf-profile-workload-partitioning.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

* For more information about creating a performance profile, see xref:../scalability_and_performance/cnf-create-performance-profiles.adoc#cnf-create-performance-profiles[Creating a performance profile].

.Sample performance profile configuration
[source,yaml]
----
include::snippets/ztp_PerformanceProfile.yaml[]
----

> include::snippets/performance-profile-workload-partitioning.adoc[]

Author: I added like this to really promote using the PPC tool as a way to create a Performance profile, I am giving this greater prominence based on dialog with QE. Keeping the snippets as reference really hand crafting a perf profile is tricky.

> include::snippets/performance-profile-workload-partitioning.adoc[]

Reviewer 3: Latest arrangement makes sense to me! 👍

When workload partitioning is enabled, the `management.workload.openshift.io/cores` resource allows the scheduler to correctly assign pods based on the `cpushares` capacity of the host, not just the default `cpuset`. This ensures more precise allocation of resources for workload partitioning scenarios.

Workload partitioning ensures that CPU requests and limits specified in the pod's configuration are respected. In {product-title} 4.16 or later, accurate CPU usage limits are set for platform pods through CPU partitioning. As workload partitioning uses the custom resource type of `management.workload.openshift.io/cores`, the values for requests and limits are the same due to a requirement by Kubernetes for extended resources. However, the annotations modified by workload partitioning correctly reflect the desired limits. 

[NOTE]
====
Extended resources cannot be overcommitted, so request and limit must be equal if both are present in a container spec.
====

include::modules/enabling-workload-partitioning.adoc[leveloffset=+1]

include::modules/create-perf-profile-workload-partitioning.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

> * For more information about creating a performance profile, see xref:../scalability_and_performance/cnf-create-performance-profiles.adoc#cnf-create-performance-profiles[Creating a performance profile].

Reviewer 4: ```suggestion
* For more information about creating a performance profile, see xref:../edge_computing/ztp-advanced-policy-config.adoc#cnf-create-performance-profiles[Creating a performance profile].
.
```

When workload partitioning is enabled, the `management.workload.openshift.io/cores` resource allows the scheduler to correctly assign pods based on the `cpushares` capacity of the host, not just the default `cpuset`. This ensures more precise allocation of resources for workload partitioning scenarios.

Workload partitioning ensures that CPU requests and limits specified in the pod's configuration are respected. In {product-title} 4.16 or later, accurate CPU usage limits are set for platform pods through CPU partitioning. As workload partitioning uses the custom resource type of `management.workload.openshift.io/cores`, the values for requests and limits are the same due to a requirement by Kubernetes for extended resources. However, the annotations modified by workload partitioning correctly reflect the desired limits. 

[NOTE]
====
Extended resources cannot be overcommitted, so request and limit must be equal if both are present in a container spec.
====

include::modules/enabling-workload-partitioning.adoc[leveloffset=+1]

include::modules/create-perf-profile-workload-partitioning.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

> * xref:../scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc#cnf-about-the-profile-creator-tool_cnf-low-latency-perf-profile[About the Performance Profile Creator].

Reviewer 3: Drop the period here:
```suggestion
* xref:../scalability_and_performance/low_latency_tuning/cnf-tuning-low-latency-nodes-with-perf-profile.adoc#cnf-about-the-profile-creator-tool_cnf-low-latency-perf-profile[About the Performance Profile Creator]
```

== Sample performance profile configuration
[source,yaml]
----
include::snippets/ztp_PerformanceProfile.yaml[]
----

include::snippets/performance-profile-workload-partitioning.adoc[]

[role="_additional-resources"]
.Additional resources

> * For the recommended workload partitioning configuration for {sno} clusters, see xref:../edge_computing/ztp-reference-cluster-configuration-for-vdu.adoc#ztp-sno-du-enabling-workload-partitioning_sno-configure-for-vdu[Recommended single-node OpenShift cluster configuration for vDU application workloads -> Workload partitioning].

Reviewer 3: Same here:
```suggestion
* For the recommended workload partitioning configuration for {sno} clusters, see xref:../edge_computing/ztp-reference-cluster-configuration-for-vdu.adoc#ztp-sno-du-enabling-workload-partitioning_sno-configure-for-vdu[Recommended single-node OpenShift cluster configuration for vDU application workloads -> Workload partitioning]
```
